themes/bootstrap3/js/check_item_statuses.js

```diff
@@ -2,153 +2,176 @@
+    //array to hold IDs and elements
+    itemStatusIds: [], itemStatusEls: [],
+    url: '/AJAX/JSON?method=getItemStatuses',
+    itemStatusRunning : false,
     dataType: 'json',
     method: 'POST',
-    url: VuFind.path + '/AJAX/JSON?method=getItemStatuses',
-    data: { 'id': itemStatusIds }
-  })
-  .done(function checkItemStatusDone(response) {
-    for (var j = 0; j < response.data.length; j++) {
-      displayItemStatus(response.data[j], itemStatusEls[response.data[j].id]);
-      itemStatusIds.splice(itemStatusIds.indexOf(response.data[j].id), 1);
-    }
-    itemStatusRunning = false;
-  })
-  .fail(function checkItemStatusFail(response, textStatus) {
-    itemStatusFail(response, textStatus);
-    itemStatusRunning = false;
-  });
-}
+    itemStatusTimer : null,
+    itemStatusDelay : 200,
 
-function itemQueueAjax(id, el) {
-  if (el.hasClass('js-item-pending')) {
-    return;
-  }
-  clearTimeout(itemStatusTimer);
-  itemStatusIds.push(id);
-  itemStatusEls[id] = el;
-  itemStatusTimer = setTimeout(runItemAjaxForQueue, itemStatusDelay);
-  el.addClass('js-item-pending').removeClass('hidden');
-  el.find('.status').removeClass('hidden');
-}
+    checkItemStatusDone: function(response) {
+        var data = response.data;
+        for (var j = 0; j < data.length; j++) {
+            displayItemStatus(data[j], this.itemStatusEls[data[j].id]);
+            this.itemStatusIds.splice(this.itemStatusIds.indexOf(data[j].id), 1);
+        }
+    },
+    itemStatusFail: function(response,textStatus) {
+        if (textStatus === 'error' ||textStatus === 'abort' || typeof response.responseJSON === 'undefined') {
+            return;
+        }
+        // display the error message on each of the ajax status place holder
+        $('.js-item-pending').addClass('text-danger').append(response.responseJSON.data);
+    },
+    itemQueueAjax: function(id, el){
+        clearTimeout(this.itemStatusTimer);
+        this.itemStatusIds.push(id);
+        this.itemStatusEls[id] = el;
+        this.itemStatusTimer = setTimeout(this.runItemAjaxForQueue.bind(this), this.itemStatusDelay);
+        el.addClass('js-item-pending').removeClass('hidden');
+        el.find('.status').removeClass('hidden');
+    },
+
+    runItemAjaxForQueue : function(){
+        if (this.itemStatusRunning) {
+            this.itemStatusTimer = setTimeout(this.runItemAjaxForQueue.bind(this), this.itemStatusDelay);
+            return;
+        }
+        $.ajax({
+            dataType: this.dataType,
+            method: this.method,
+            url: VuFind.path+this.url,
+            context: this,
+            data: { 'id': this.itemStatusIds }
+        })
+            .done(this.checkItemStatusDone)
+            .fail( this.itemStatusFail)
+            .always(function(){
+                this.itemStatusRunning = false;
+            });
+    }//end runItemAjax
+};
+
+//store the handlers in a "hash" obj
+//add you own overridden handler here
+var handlers = {
```

Comment:
Could we encapsulate this or give it a more specific name? I'd imagine `handlers` could be easily and accidentally overwritten by local custom JS.

Reply:
It's a good idea.  I'll defer to you on which to do.